langroid/agent/special/sql_chat_agent.py

```diff
@@ -0,0 +1,293 @@
+    debug: bool = False
+    stream: bool = True  # allow streaming where needed
+    database_uri: str  # Database URI
+    retry_query: int = 3  # Number of times to retry query
+    vecdb: None | VectorStoreConfig = None
+    context_descriptions: Optional[
+        Dict[str, Dict[str, Union[str, Dict[str, str]]]]
+    ] = None
+
+    """
+    Optional, but strongly recommended, context descriptions for tables, columns, 
+    and relationships. It should be a dictionary where each key is a table name 
+    and its value is another dictionary. 
+
+    In this inner dictionary:
+    - The 'description' key corresponds to a string description of the table.
+    - The 'columns' key corresponds to another dictionary where each key is a 
+    column name and its value is a string description of that column.
+    - The 'relationships' key corresponds to another dictionary where each key 
+    is another table name and the value is a description of the relationship to 
+    that table.
+
+    For example:
+    {
+        'table1': {
+            'description': 'description of table1',
+            'columns': {
+                'column1': 'description of column1 in table1',
+                'column2': 'description of column2 in table1'
+            }
+        },
+        'table2': {
+            'description': 'description of table2',
+            'columns': {
+                'column3': 'description of column3 in table2',
+                'column4': 'description of column4 in table2'
+            }
+        }
+    }
+    """
+
+    llm: OpenAIGPTConfig = OpenAIGPTConfig(
+        type="openai",
+        chat_model=OpenAIChatModel.GPT4,
+        completion_model=OpenAIChatModel.GPT4,
+    )
+    prompts: PromptsConfig = PromptsConfig(
+        max_tokens=1000,
+    )
+
+
+class RunQueryTool(ToolMessage):
+    request: str = "run_query"
+    purpose: str = """
+            To run <query> on the database 'db' and 
+            return the results to answer a question.
+            """
+    query: str
+
+
+class SQLChatAgent(ChatAgent):
+    """
+    Agent for chatting with a collection of documents.
+    """
+
+    def __init__(self, config: SQLChatAgentConfig):
+        super().__init__(config)
+        self.config: SQLChatAgentConfig = config
+        self.engine = create_engine(config.database_uri)
+        self.Session = sessionmaker(bind=self.engine)
+        self.metadata = MetaData()
+        self.metadata.reflect(self.engine)
+        self.tables_info = ""
+
+        # User-provided context descriptions
+        context_descriptions = config.context_descriptions
+
+        logger.info(
+            f"""SQLChatAgent initialized with database: 
+            {self.engine} and tables: 
+            {self.metadata.tables}
+            """
+        )
+
+        # Create inspector
+        inspector = inspect(self.engine)
+
+        # Gather table and column information
+        self.tables_info = self.gather_table_info(context_descriptions, inspector)
```

Comment:
As I understand it, `context_descriptions` are in JSON format, and `gather_table_info` is massaging this into a more human-readable format, and then this is inserted into the LLM context. But do we need to do this? We could directly insert the `context_descriptions` into the prompt, and add a short explanation of the structure of the JSON.